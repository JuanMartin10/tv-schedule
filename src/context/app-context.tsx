import { createContext, useContext, useEffect, useState } from 'react';
import { type Channel } from '../models/api-types';
import { getTimeFromDate } from '../util/functions';
import { useChannels } from '../hooks/useChannels';

export interface AppContextTypes {
  loading: boolean;
  channels: Channel[];
  currentTime: number;
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
  setChannels: React.Dispatch<React.SetStateAction<Channel[]>>;
}

interface ContextProps {
  children: React.ReactNode;
}

const AppContext = createContext<AppContextTypes | undefined>(undefined);

export const AppProvider: React.FC<ContextProps> = ({ children }) => {
  const [loading, setLoading] = useState(false);
  const [currentTime, setCurrentTime] = useState<number>(0);
  const { channels, setChannels } = useChannels({ setLoading });

  const currentDate = new Date();

  useEffect(() => {
    setCurrentTime(getTimeFromDate(currentDate));
  }, []);

  return (
    <AppContext.Provider
      value={{
        loading,
        channels,
        currentTime,
        setLoading,
        setChannels,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = (): AppContextTypes => {
  const context = useContext(AppContext);
  if (context == null) {
    throw new Error('useAppContext must be used within an AppWrapper');
  }
  return context;
};
