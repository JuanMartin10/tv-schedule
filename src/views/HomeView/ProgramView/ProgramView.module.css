* {
  margin: 0;
}

.imageContainer {
  position: relative;
}
.image {
  width: 100%;
}
.live {
  background-color: rgba(0, 0, 0, 0.5);
  color: #fff;
  padding: 4px 8px;
  position: absolute;
  top: 20px;
  right: 20px;
  z-index: 100;
  border-radius: 4px;
  padding-left: 25px;
}

.playButton {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 80px;
  height: 80px;
  border-radius: 50%;
  background-color: rgba(255, 255, 255, 0.8);
  background-image: url('https://static.thenounproject.com/png/1192912-200.png');
  background-repeat: no-repeat;
  background-size: 60%;
  background-position: center;
  cursor: pointer;
}
.live::before {
  content: '';
  position: absolute;
  top: 50%;
  left: 10px;
  transform: translateY(-50%);
  width: 10px;
  height: 10px;
  border-radius: 50%;
  background-color: red;
  animation: blink 1s ease-in-out infinite;
}
.programView {
  display: flex;
  flex-direction: column;
  gap: 8px;
  flex: 1;
  margin-bottom: 150px;
  padding: 0 10px;
}
.info {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  gap: 8px;
  margin: 0 4px;
}
.infoTitle {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
  gap: 4px;
}
.channel,
.infoDetails {
  display: flex;
  gap: 8px;
  align-items: center;
  justify-content: center;
  font-size: 14px;
}

.title {
  font-size: 30px;
}

.day::before {
  content: '';
  display: inline-block;
  width: 5px;
  height: 5px;
  background-color: #808080;
  border-radius: 50%;
  margin-right: 5px;
  margin-bottom: 2px;
}

.description {
  display: flex;
  flex-direction: column;
  gap: 24px;
  margin: 0 4px;
}
.grey {
  color: #808080;
}

.white {
  color: #fff;
}

@media (min-width: 600px) {
  .programView {
    padding: 0;
    margin-bottom: 50px;
    gap: 32px;
  }
  .info {
    gap: 16px;
  }
  .channel,
  .infoDetails {
    font-size: 16px;
  }
}

@keyframes blink {
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0.5;
  }
  100% {
    opacity: 1;
  }
}
