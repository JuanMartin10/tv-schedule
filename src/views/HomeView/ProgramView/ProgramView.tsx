import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import { useAppContext } from '@context/app-context';
import Icon from '@components/Icon/Icon';
import { MOCK_PROGRAM_IMAGE } from '@mocks';
import {
  formatDate,
  formatHours,
  isProgramActive,
  getTimeFromDateISO,
} from '@util/functions';
import styles from './ProgramView.module.css';

const ProgramView = () => {
  const {
    state: { program, channel },
  } = useLocation();
  const { currentTime } = useAppContext();

  const [isLive, setIsLive] = useState(false);

  useEffect(() => {
    const res = isProgramActive(
      getTimeFromDateISO(program.start),
      getTimeFromDateISO(program.end),
      currentTime
    );
    setIsLive(res);
  }, []);

  return (
    <div className={styles.programView}>
      <section className={styles.imageContainer}>
        <img src={MOCK_PROGRAM_IMAGE} className={styles.image} />
        {isLive && <div className={styles.playButton} />}
        {isLive && <div className={styles.live}>Live</div>}
      </section>
      <section className={styles.info}>
        <div>
          <img src={channel.images.logo} width={100} />
        </div>
        <div className={styles.infoTitle}>
          <div className={`${styles.channel} ${styles.grey}`}>
            <h5 className={styles.white}>{channel.title}</h5>
            <span>{`${formatHours(program.start)}-${formatHours(
              program.end
            )}`}</span>
            <span className={styles.day}> {formatDate(program.start)} </span>
          </div>
          <h3 className={styles.title}>{program.title}</h3>
          <div className={`${styles.infoDetails} ${styles.grey}`}>
            <span>2015</span>
            <span>Romance</span>
            <span>Drama</span>
            <span>Horror</span>
          </div>
        </div>
        <div>
          <Icon icon='ClockIcon' color='white' variant='solid' />
        </div>
      </section>
      <div className={styles.description}>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
          eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
          minim veniam, quis nostrud exercitation ullamco laboris nisi ut
          aliquip ex ea commodo consequat. Duis aute irure dolor in
          reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
          pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
          culpa qui officia deserunt mollit anim id est laborum
        </p>
        <div>
          <p className={styles.grey}>
            Cast: Wagner Moura, Boy Holbruk, Pedro Pascal
          </p>
          <p className={styles.grey}>
            Creators: Wagner Moura, Boy Holbruk, Pedro Pascal
          </p>
        </div>
      </div>
    </div>
  );
};

export default ProgramView;
