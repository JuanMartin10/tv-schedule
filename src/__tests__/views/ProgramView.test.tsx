import { AppContext } from '@context/app-context';
import { cleanup, render, waitFor, screen } from '@testing-library/react';
import channelsJson from '@mocks/data.json';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import ProgramView from '../../views/HomeView/ProgramView/ProgramView';

const renderWithProvider = (ui: React.ReactElement) =>
  render(
    <AppContext.Provider
      value={{
        channels: channelsJson.channels,
        loading: false,
        currentTime: 0,
        startTime: 0,
        endTime: 0,
        timesArray: [],
        bookmarkPosition: 0,
        setLoading: () => {},
        setChannels: () => {},
      }}
    >
      {ui}
    </AppContext.Provider>
  );

const channel = channelsJson.channels[1];
const program = channelsJson.channels[1].schedules[1];

const wrapper = ({ channelId, programId, children }: any) => (
  <MemoryRouter
    initialEntries={[
      `/${channelId}/${programId}`,
      { state: { program, channel } },
    ]}
  >
    <Routes>
      <Route path='/channelId/:programId' Component={ProgramView}></Route>
    </Routes>
  </MemoryRouter>
);
describe('<ProgramView />', () => {
  const wrapperId = () =>
    wrapper({
      channelId: 'sky1',
      programId: 'dummy_program_id',
    });

  beforeEach(() => {
    renderWithProvider(wrapperId());
  });

  afterEach(cleanup);

  it('Should render', () => {
    renderWithProvider(wrapperId());
  });

  it('Should match the snapshot', () => {
    const { container } = renderWithProvider(wrapperId());
    expect(container).toMatchSnapshot();
  });

  it('Should render podcast image', () => {
    waitFor(() => {
      const images = screen.getAllByRole('img');
      expect(images).toHaveLength(1);
      expect(images).toBeDefined();
    });
  });

  it('Should render program description', () => {
    waitFor(() => {
      const description = screen.getByText(
        /Lorem ipsum dolor sit amet, consectetur adip/i
      );
      expect(description).toBeDefined();
    });
  });
});
