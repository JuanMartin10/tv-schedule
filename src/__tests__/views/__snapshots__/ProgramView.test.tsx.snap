// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ProgramView /> > Should match the snapshot 1`] = `<div />`;
